test/mocks/buffer/mocks.h

```diff
@@ -9,24 +10,22 @@
 
 namespace Envoy {
 
-class MockBuffer : public Buffer::OwnedImpl {
+// A template class to allow code reuse between MockBuffer and MockWatermarkBuffer
```

Comment:
I don't have a strong opinion here, but FWIW I've avoided doing things like this in the past and just duplicated all the code (see for example MockClientConnection/MockConnection/MockConnectionBase). This is one area where I guess I just live with the duplication because the template code is kind of crazy as you point out. I'm fine either way.